sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/utils/EdsProtocolTransformer.java

```diff
@@ -16,112 +16,104 @@
-            if (!serviceNameFromCluster.isPresent()) {
-                continue;
-            }
-            String serviceName = serviceNameFromCluster.get();
-            nameMapping.computeIfAbsent(serviceName, key -> new HashSet<>()).add(cluster.getName());
-        }
-        return nameMapping;
+    private EdsProtocolTransformer() {
     }
 
     /**
-     * get the instance of one service by xds protocol
+     * get service instances by xds protocol
      *
      * @param loadAssignments eds data
-     * @return instance of service
+     * @return instances of service
      */
-    public static Set<ServiceInstance> getServiceInstances(
+    public static XdsServiceClusterLoadAssigment getServiceInstances(
             List<ClusterLoadAssignment> loadAssignments) {
-        return loadAssignments.stream()
+        XdsServiceClusterLoadAssigment serviceClusterLoadAssigment = new XdsServiceClusterLoadAssigment();
+
+        Map<String, XdsClusterLoadAssigment> clusterLoadAssigmentMap = loadAssignments.stream()
                 .filter(Objects::nonNull)
-                .flatMap(loadAssignment -> getServiceInstancesFromLoadAssignment(loadAssignment).stream())
-                .collect(Collectors.toSet());
+                .map(loadAssignment -> parseClusterLoadAssignment(loadAssignment))
+                .filter(clusterInstance -> clusterInstance.getClusterName() != null)
+                .collect(Collectors.toMap(XdsClusterLoadAssigment::getClusterName,
+                        clusterInstance -> clusterInstance));
+        serviceClusterLoadAssigment.setClusterLoadAssigments(clusterLoadAssigmentMap);
+        serviceClusterLoadAssigment.setBaseClusterName(getServiceBaseClusterName(clusterLoadAssigmentMap));
+        return serviceClusterLoadAssigment;
     }
 
-    private static Set<ServiceInstance> getServiceInstancesFromLoadAssignment(ClusterLoadAssignment loadAssignment) {
+    private static XdsClusterLoadAssigment parseClusterLoadAssignment(ClusterLoadAssignment loadAssignment) {
         String clusterName = loadAssignment.getClusterName();
-        Optional<String> serviceNameOptional = getServiceNameFromCluster(clusterName);
+        Optional<String> serviceNameOptional = XdsCommonUtils.getServiceNameFromCluster(clusterName);
+        XdsClusterLoadAssigment xdsClusterLoadAssigment = new XdsClusterLoadAssigment();
         if (!serviceNameOptional.isPresent()) {
-            return Collections.EMPTY_SET;
+            return xdsClusterLoadAssigment;
         }
         String serviceName = serviceNameOptional.get();
-        return processClusterLoadAssignment(loadAssignment, serviceName, clusterName);
+        xdsClusterLoadAssigment.setClusterName(clusterName);
+        xdsClusterLoadAssigment.setServiceName(serviceName);
+        xdsClusterLoadAssigment
+                .setLocalityInstances(parseLocalityLbEndpointsList(loadAssignment, serviceName, clusterName));
+        return xdsClusterLoadAssigment;
```

Comment:
define consurctors for XdsClusterLoadAssigment and XdsServiceClusterLoadAssigment to avoid these set methods since you are create objects not update them

Reply:
done